src/resources/listing.js

```diff
@@ -0,0 +1,225 @@
+//
+class Listing {
+  /**
+   * A Listing is constructed based on its on-chain and off-chain data.
+   * @param {Object} chainListing -
+   * @param {Object} ipfsListing
+   */
+  constructor(chainListing, ipfsListing) {
+    // FIXME(franck): Exposing directly the chain data will make it difficult
+    // to support backward compatibility of the Listing interface in the future. We should
+    // select and possibly abstract what data from the chain gets exposed.
+    Object.assign(this, chainListing, ipfsListing)
+  }
+
+  get unitsSold() {
+    // Lazy caching.
+    if (this._unitsSold !== undefined) {
+      return this._unitsSold
+    }
+    this._unitsSold = Object.keys(this.offers).reduce((acc, offerId) => {
+      if (this.offers[offerId].status === 'created') {
+        return acc + 1
+      }
+      // TODO: we need to subtract 1 for every offer that is canceled
+      return acc
+    }, 0)
+    return this._unitsSold
+  }
+
+  get unitsAvailable() {
+    // Units available is derived from units for sale and offers created.
+    // Should never be negative.
+    return Math.max(this.unitsForSale - this.unitsSold, 0)
+  }
+}
+
+/**
+ * Returns an adapter based on a schema version.
+ * @param {string} schemaVersion
+ * @returns {ListingSchemaAdapter}
+ * @throws {Error}
+ */
+function schemaAdapterFactory(schemaVersion) {
+  if (schemaVersion === SCHEMA_V1) {
+    return SchemaAdapterV1
+  } else {
+    throw new Error(`Unsupported schema version ${schemaVersion}`)
+  }
+}
+
+class SchemaAdapterV1 {
+  /**
+   * Validates the data is compliant with Origin Protocol core listing schema V1.
+   * @param {object} data - Listing data to validate.
+   * @throws {Error}
+   */
+  static validate(data) {
+    if (data.schemaVersion !== SCHEMA_V1) {
+      throw new Error(`Listing schema version ${data.schemaVersion} != ${SCHEMA_V1}`)
+    }
+
+    const validator = ajv.getSchema(schemaIdV1)
+    if (!validator(data)) {
+      throw new Error(
+        `Invalid V1 listing data.
+        Listing: ${JSON.stringify(data)}.
+        Schema: ${JSON.stringify(schemaV1)}
+        Validation errors: ${JSON.stringify(validator.errors)}`
+      )
+    }
+  }
+
+  /**
+   * Populates an IpfsListing object from listing data that uses V1 schema.
+   * @param {object} data - Listing data, expected to use schema V1.
+   * @throws {Error}
+   */
+  static deserialize(data) {
+    // Validate the data against the Origin Protocol V1 core listing schema.
+    SchemaAdapterV1.validate(data)
+
+    const listing = {
+      schemaVersion:  data.schemaVersion,
+      type:           data.listingType,
+      category:       data.category,
+      subCategory:    data.subCategory,
+      language:       data.language,
+      title:          data.title,
+      description:    data.description,
+      media:          data.media,
+      expiry:         data.expiry,
+    }
+
+    // Unit data.
+    if (listing.type === 'unit') {
```

Comment:
Is it worth treating unit and fractional listing types as fundamentally different now that they're the same on-chain?

Reply:
They are identical onchain but differ off-chain in the IPFS data. This method handles deserializing the data from off-chain, that's why unit vs fractional are handled differently.

Now, you are right that for our V1 release we won't support fractional. But I thought would be good to have it here as a placeholder. Since fractional usage schema is purely additive, I was thinking that when we finalize it after our V1 release, we can just put it in schema V1. We won't have to put in a schema V2, which will save us from writing a new SchemaAdapterV2.

Having said that, if this is too confusing, I can definitely remove any reference to fractional for now.

Reply:
Yeah fractional listings differ enough in ipfs data format that I think it makes sense to make them separate types. Fractional listings have the concept of time slots and the restriction that multiple bookings for the same item should not overlap.